File: airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/DefaultNormalizationRunner.java
/*
 * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.normalization;

import static io.airbyte.workers.process.Metadata.JOB_TYPE_KEY;
import static io.airbyte.workers.process.Metadata.NORMALIZE_STEP;
import static io.airbyte.workers.process.Metadata.SYNC_JOB;
import static io.airbyte.workers.process.Metadata.SYNC_STEP_KEY;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.common.base.Strings;
import com.google.common.collect.ImmutableMap;
import io.airbyte.commons.io.IOs;
import io.airbyte.commons.io.LineGobbler;
import io.airbyte.commons.json.Jsons;
import io.airbyte.commons.logging.LoggingHelper.Color;
import io.airbyte.commons.logging.MdcScope;
import io.airbyte.commons.logging.MdcScope.Builder;
import io.airbyte.config.OperatorDbt;
import io.airbyte.config.ResourceRequirements;
import io.airbyte.persistence.job.errorreporter.SentryExceptionHelper;
import io.airbyte.protocol.models.AirbyteErrorTraceMessage;
import io.airbyte.protocol.models.AirbyteErrorTraceMessage.FailureType;
import io.airbyte.protocol.models.AirbyteMessage;
import io.airbyte.protocol.models.AirbyteMessage.Type;
import io.airbyte.protocol.models.AirbyteTraceMessage;
import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
import io.airbyte.workers.WorkerConstants;
import io.airbyte.workers.WorkerUtils;
import io.airbyte.workers.exception.WorkerException;
import io.airbyte.workers.process.ProcessFactory;
import java.io.InputStream;
import java.nio.file.Path;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class DefaultNormalizationRunner implements NormalizationRunner {

  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultNormalizationRunner.class);
  private static final MdcScope.Builder CONTAINER_LOG_MDC_BUILDER = new Builder()
      .setLogPrefix("normalization")
      .setPrefixColor(Color.GREEN_BACKGROUND);

  private final String normalizationIntegrationType;
  private final ProcessFactory processFactory;
  private final String normalizationImageName;
  private final NormalizationAirbyteStreamFactory streamFactory = new NormalizationAirbyteStreamFactory(CONTAINER_LOG_MDC_BUILDER);
  private Map<Type, List<AirbyteMessage>> airbyteMessagesByType;
  private String dbtErrorStack;

  private Process process = null;

  public DefaultNormalizationRunner(final ProcessFactory processFactory,
                                    final String normalizationImage,
                                    final String normalizationIntegrationType) {
    this.processFactory = processFactory;
    this.normalizationImageName = normalizationImage;
    this.normalizationIntegrationType = normalizationIntegrationType;
  }

  @Override
  public boolean configureDbt(final String jobId,
                              final int attempt,
                              final Path jobRoot,
                              final JsonNode config,
                              final ResourceRequirements resourceRequirements,
                              final OperatorDbt dbtConfig)
      throws Exception {
    final Map<String, String> files = ImmutableMap.of(
        WorkerConstants.DESTINATION_CONFIG_JSON_FILENAME, Jsons.serialize(config));
    final String gitRepoUrl = dbtConfig.getGitRepoUrl();
    if (Strings.isNullOrEmpty(gitRepoUrl)) {
      throw new WorkerException("Git Repo Url is required");
    }
    final String gitRepoBranch = dbtConfig.getGitRepoBranch();
    if (Strings.isNullOrEmpty(gitRepoBranch)) {
      return runProcess(jobId, attempt, jobRoot, files, resourceRequirements, "configure-dbt",
          "--integration-type", normalizationIntegrationType.toLowerCase(),
          "--config", WorkerConstants.DESTINATION_CONFIG_JSON_FILENAME,
          "--git-repo", gitRepoUrl);
    } else {
      return runProcess(jobId, attempt, jobRoot, files, resourceRequirements, "configure-dbt",
          "--integration-type", normalizationIntegrationType.toLowerCase(),
          "--config", WorkerConstants.DESTINATION_CONFIG_JSON_FILENAME,
          "--git-repo", gitRepoUrl,
          "--git-branch", gitRepoBranch);
    }
  }

  @Override
  public boolean normalize(final String jobId,
                           final int attempt,
                           final Path jobRoot,
                           final JsonNode config,
                           final ConfiguredAirbyteCatalog catalog,
                           final ResourceRequirements resourceRequirements)
      throws Exception {
    final Map<String, String> files = ImmutableMap.of(
        WorkerConstants.DESTINATION_CONFIG_JSON_FILENAME, Jsons.serialize(config),
        WorkerConstants.DESTINATION_CATALOG_JSON_FILENAME, Jsons.serialize(catalog));

    return runProcess(jobId, attempt, jobRoot, files, resourceRequirements, "run",
        "--integration-type", normalizationIntegrationType.toLowerCase(),
        "--config", WorkerConstants.DESTINATION_CONFIG_JSON_FILENAME,
        "--catalog", WorkerConstants.DESTINATION_CATALOG_JSON_FILENAME);
  }

  @SuppressWarnings("PMD.AvoidLiteralsInIfCondition")
  private boolean runProcess(final String jobId,
                             final int attempt,
                             final Path jobRoot,
                             final Map<String, String> files,
                             final ResourceRequirements resourceRequirements,
                             final String... args)
      throws Exception {
    try {
      LOGGER.info("Running with normalization version: {}", normalizationImageName);
      process = processFactory.create(
          NORMALIZE_STEP,
          jobId,
          attempt,
          jobRoot,
          normalizationImageName,
          // custom connector does not use normalization
          false,
          false, files,
          null,
          resourceRequirements,
          null,
          Map.of(JOB_TYPE_KEY, SYNC_JOB, SYNC_STEP_KEY, NORMALIZE_STEP),
          Collections.emptyMap(),
          Collections.emptyMap(),
          args);

      try (final InputStream stdout = process.getInputStream()) {
        // finds and collects any AirbyteMessages from stdout
        // also builds a list of raw dbt errors and stores in streamFactory
        airbyteMessagesByType = streamFactory.create(IOs.newBufferedReader(stdout))
            .collect(Collectors.groupingBy(AirbyteMessage::getType));

        // picks up error logs from dbt
        dbtErrorStack = String.join("\n", streamFactory.getDbtErrors());

        if (!"".equals(dbtErrorStack)) {
          final AirbyteMessage dbtTraceMessage = new AirbyteMessage()
              .withType(Type.TRACE)
              .withTrace(new AirbyteTraceMessage()
                  .withType(AirbyteTraceMessage.Type.ERROR)
                  .withEmittedAt((double) System.currentTimeMillis())
                  .withError(new AirbyteErrorTraceMessage()
                      .withFailureType(FailureType.SYSTEM_ERROR) // TODO: decide on best FailureType for this
                      .withMessage("Normalization failed during the dbt run. This may indicate a problem with the data itself.")
                      .withInternalMessage(buildInternalErrorMessageFromDbtStackTrace())
                      // due to the lack of consistent defining features in dbt errors we're injecting a breadcrumb to the
                      // stacktrace so we can confidently identify all dbt errors when parsing and sending to Sentry
                      // see dbt error examples: https://docs.getdbt.com/guides/legacy/debugging-errors for more context
                      .withStackTrace("AirbyteDbtError: \n".concat(dbtErrorStack))));

          airbyteMessagesByType.putIfAbsent(Type.TRACE, List.of(dbtTraceMessage));
        }
      }
      LineGobbler.gobble(process.getErrorStream(), LOGGER::error, CONTAINER_LOG_MDC_BUILDER);

      WorkerUtils.wait(process);

      return process.exitValue() == 0;
    } catch (final Exception e) {
      // make sure we kill the process on failure to avoid zombies.
      if (process != null) {
        WorkerUtils.cancelProcess(process);
      }
      throw e;
    }
  }

  @Override
  public void close() throws Exception {
    if (process == null) {
      return;
    }

    LOGGER.info("Terminating normalization process...");
    WorkerUtils.gentleClose(process, 1, TimeUnit.MINUTES);

    /*
     * After attempting to close the process check the following:
     *
     * Did the process actually terminate? If "yes", did it do so nominally?
     */
    if (process.isAlive()) {
      throw new WorkerException("Normalization process did not terminate after 1 minute.");
    } else if (process.exitValue() != 0) {
      throw new WorkerException("Normalization process did not terminate normally (exit code: " + process.exitValue() + ")");
    } else {
      LOGGER.info("Normalization process successfully terminated.");
    }
  }

  @Override
  public Stream<AirbyteTraceMessage> getTraceMessages() {
    if (airbyteMessagesByType != null && airbyteMessagesByType.get(Type.TRACE) != null) {
      return airbyteMessagesByType.get(Type.TRACE).stream().map(AirbyteMessage::getTrace);
    }
    return Stream.empty();
  }

  private String buildInternalErrorMessageFromDbtStackTrace() {
    final Map<SentryExceptionHelper.ERROR_MAP_KEYS, String> errorMap = SentryExceptionHelper.getUsefulErrorMessageAndTypeFromDbtError(dbtErrorStack);
    return errorMap.get(SentryExceptionHelper.ERROR_MAP_KEYS.ERROR_MAP_MESSAGE_KEY);
  }

}

File: airbyte-cdk/java/airbyte-cdk/dependencies/src/testFixtures/kotlin/io/airbyte/workers/normalization/DefaultNormalizationRunner.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.workers.normalization

import com.fasterxml.jackson.databind.JsonNode
import com.google.common.base.Strings
import com.google.common.collect.ImmutableMap
import io.airbyte.commons.io.IOs
import io.airbyte.commons.io.LineGobbler
import io.airbyte.commons.json.Jsons
import io.airbyte.commons.logging.LoggingHelper
import io.airbyte.commons.logging.MdcScope
import io.airbyte.configoss.OperatorDbt
import io.airbyte.configoss.ResourceRequirements
import io.airbyte.protocol.models.AirbyteErrorTraceMessage
import io.airbyte.protocol.models.AirbyteMessage
import io.airbyte.protocol.models.AirbyteTraceMessage
import io.airbyte.protocol.models.ConfiguredAirbyteCatalog
import io.airbyte.workers.TestHarnessUtils
import io.airbyte.workers.WorkerConstants
import io.airbyte.workers.exception.TestHarnessException
import io.airbyte.workers.process.Metadata
import io.airbyte.workers.process.ProcessFactory
import io.github.oshai.kotlinlogging.KotlinLogging
import java.nio.file.Path
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.function.Function
import java.util.stream.Collectors
import java.util.stream.Stream

private val LOGGER = KotlinLogging.logger {}

class DefaultNormalizationRunner(
    private val processFactory: ProcessFactory,
    private val normalizationImageName: String?,
    private val normalizationIntegrationType: String?
) : NormalizationRunner {
    private val streamFactory = NormalizationAirbyteStreamFactory(CONTAINER_LOG_MDC_BUILDER)
    private var airbyteMessagesByType: MutableMap<AirbyteMessage.Type, List<AirbyteMessage>> =
        HashMap()
    private var dbtErrorStack: String? = null

    private var process: Process? = null

    @Throws(Exception::class)
    override fun configureDbt(
        jobId: String,
        attempt: Int,
        jobRoot: Path,
        config: JsonNode?,
        resourceRequirements: ResourceRequirements?,
        dbtConfig: OperatorDbt
    ): Boolean {
        val files: Map<String?, String?> =
            ImmutableMap.of(
                WorkerConstants.DESTINATION_CONFIG_JSON_FILENAME,
                Jsons.serialize(config)
            )
        val gitRepoUrl = dbtConfig.gitRepoUrl
        if (Strings.isNullOrEmpty(gitRepoUrl)) {
            throw TestHarnessException("Git Repo Url is required")
        }
        val gitRepoBranch = dbtConfig.gitRepoBranch
        return if (Strings.isNullOrEmpty(gitRepoBranch)) {
            runProcess(
                jobId,
                attempt,
                jobRoot,
                files,
                resourceRequirements,
                "configure-dbt",
                "--integration-type",
                normalizationIntegrationType!!.lowercase(Locale.getDefault()),
                "--config",
                WorkerConstants.DESTINATION_CONFIG_JSON_FILENAME,
                "--git-repo",
                gitRepoUrl
            )
        } else {
            runProcess(
                jobId,
                attempt,
                jobRoot,
                files,
                resourceRequirements,
                "configure-dbt",
                "--integration-type",
                normalizationIntegrationType!!.lowercase(Locale.getDefault()),
                "--config",
                WorkerConstants.DESTINATION_CONFIG_JSON_FILENAME,
                "--git-repo",
                gitRepoUrl,
                "--git-branch",
                gitRepoBranch
            )
        }
    }

    @Throws(Exception::class)
    override fun normalize(
        jobId: String,
        attempt: Int,
        jobRoot: Path,
        config: JsonNode,
        catalog: ConfiguredAirbyteCatalog,
        resourceRequirements: ResourceRequirements?
    ): Boolean {
        val files: Map<String?, String?> =
            ImmutableMap.of(
                WorkerConstants.DESTINATION_CONFIG_JSON_FILENAME,
                Jsons.serialize(config),
                WorkerConstants.DESTINATION_CATALOG_JSON_FILENAME,
                Jsons.serialize(catalog)
            )

        return runProcess(
            jobId,
            attempt,
            jobRoot,
            files,
            resourceRequirements,
            "run",
            "--integration-type",
            normalizationIntegrationType!!.lowercase(Locale.getDefault()),
            "--config",
            WorkerConstants.DESTINATION_CONFIG_JSON_FILENAME,
            "--catalog",
            WorkerConstants.DESTINATION_CATALOG_JSON_FILENAME
        )
    }

    @Throws(Exception::class)
    private fun runProcess(
        jobId: String,
        attempt: Int,
        jobRoot: Path,
        files: Map<String?, String?>,
        resourceRequirements: ResourceRequirements?,
        vararg args: String
    ): Boolean {
        try {
            LOGGER.info("Running with normalization version: {}", normalizationImageName)
            var process =
                processFactory.create(
                    Metadata.NORMALIZE_STEP,
                    jobId,
                    attempt,
                    jobRoot,
                    normalizationImageName!!, // custom connector does not use normalization
                    false,
                    false,
                    files,
                    null,
                    resourceRequirements,
                    null,
                    java.util.Map.of(
                        Metadata.JOB_TYPE_KEY,
                        Metadata.SYNC_JOB,
                        Metadata.SYNC_STEP_KEY,
                        Metadata.NORMALIZE_STEP
                    ),
                    emptyMap(),
                    emptyMap(),
                    emptyMap(),
                    *args
                )
            this.process = process

            process.inputStream.use { stdout ->
                // finds and collects any AirbyteMessages from stdout
                // also builds a list of raw dbt errors and stores in streamFactory
                airbyteMessagesByType =
                    streamFactory
                        .create(IOs.newBufferedReader(stdout))
                        .collect(
                            Collectors.groupingBy(Function { obj: AirbyteMessage -> obj.type })
                        )

                // picks up error logs from dbt
                dbtErrorStack = java.lang.String.join("\n", streamFactory.dbtErrors)
                if ("" != dbtErrorStack) {
                    val dbtTraceMessage =
                        AirbyteMessage()
                            .withType(AirbyteMessage.Type.TRACE)
                            .withTrace(
                                AirbyteTraceMessage()
                                    .withType(AirbyteTraceMessage.Type.ERROR)
                                    .withEmittedAt(System.currentTimeMillis().toDouble())
                                    .withError(
                                        AirbyteErrorTraceMessage()
                                            .withFailureType(
                                                AirbyteErrorTraceMessage.FailureType.SYSTEM_ERROR
                                            ) // TODO: decide on best FailureType for this
                                            .withMessage(
                                                "Normalization failed during the dbt run. This may indicate a problem with the data itself."
                                            ) // due to the lack of consistent defining features in
                                            // dbt errors we're injecting a breadcrumb to the
                                            // stacktrace so we can confidently identify all dbt
                                            // errors when parsing and sending to Sentry
                                            // see dbt error examples:
                                            // https://docs.getdbt.com/guides/legacy/debugging-errors for more context
                                            .withStackTrace("AirbyteDbtError: \n$dbtErrorStack")
                                    )
                            )

                    airbyteMessagesByType.putIfAbsent(
                        AirbyteMessage.Type.TRACE,
                        java.util.List.of(dbtTraceMessage)
                    )
                }
            }
            LineGobbler.gobble(
                process.errorStream,
                { msg: String -> LOGGER.error(msg) },
                CONTAINER_LOG_MDC_BUILDER
            )

            TestHarnessUtils.wait(process)

            return process.exitValue() == 0
        } catch (e: Exception) {
            // make sure we kill the process on failure to avoid zombies.
            process?.let { TestHarnessUtils.cancelProcess(process) }
            throw e
        }
    }

    @Throws(Exception::class)
    override fun close() {
        process?.let {
            LOGGER.info("Terminating normalization process...")
            TestHarnessUtils.gentleClose(it, 1, TimeUnit.MINUTES)

            /*
             * After attempting to close the process check the following:
             *
             * Did the process actually terminate? If "yes", did it do so nominally?
             */
            if (it.isAlive) {
                throw TestHarnessException(
                    "Normalization process did not terminate after 1 minute."
                )
            } else if (it.exitValue() != 0) {
                throw TestHarnessException(
                    "Normalization process did not terminate normally (exit code: " +
                        it.exitValue() +
                        ")"
                )
            } else {
                LOGGER.info("Normalization process successfully terminated.")
            }
        }
    }

    override val traceMessages: Stream<AirbyteTraceMessage>
        get() {
            if (airbyteMessagesByType[AirbyteMessage.Type.TRACE] != null) {
                return airbyteMessagesByType[AirbyteMessage.Type.TRACE]!!
                    .map { obj: AirbyteMessage -> obj.trace }
                    .stream()
            }
            return Stream.empty()
        }

    companion object {

        private val CONTAINER_LOG_MDC_BUILDER: MdcScope.Builder =
            MdcScope.Builder()
                .setLogPrefix("normalization")
                .setPrefixColor(LoggingHelper.Color.GREEN_BACKGROUND)
    }
}
